#include "logindialog_ui.h"
#include "ui_logindialog.h"
#include "registerdialog_ui.h"
#include <QDebug>

LoginDialog::LoginDialog(QWidget *parent)
    : QDialog(parent)
    , ui(new Ui::LoginDialog)
{
    ui->setupUi(this);
    setupConnections();
    
    // Create default admin account if users.txt doesn't exist
    QFile file("users.txt");
    if (!file.exists()) {
        saveUserCredentials("admin", "admin123");
        saveUserCredentials("user", "password");
    }
    
    ui->usernameEdit->setFocus();
}

LoginDialog::~LoginDialog()
{
    delete ui;
}

void LoginDialog::setupConnections()
{
    // Connect button signals
    connect(ui->loginButton, &QPushButton::clicked, this, &LoginDialog::onLoginClicked);
    connect(ui->resetButton, &QPushButton::clicked, this, &LoginDialog::onResetClicked);
    connect(ui->registerButton, &QPushButton::clicked, this, &LoginDialog::onRegisterClicked);
    connect(ui->togglePasswordButton, &QPushButton::clicked, this, &LoginDialog::togglePasswordVisibility);
    
    // Enter key support
    connect(ui->usernameEdit, &QLineEdit::returnPressed, this, &LoginDialog::onLoginClicked);
    connect(ui->passwordEdit, &QLineEdit::returnPressed, this, &LoginDialog::onLoginClicked);
}

void LoginDialog::onLoginClicked()
{
    QString username = ui->usernameEdit->text().trimmed();
    QString password = ui->passwordEdit->text();
    
    if (username.isEmpty()) {
        QMessageBox::warning(this, "ÁôªÂΩïÂ§±Ë¥•", "ËØ∑ËæìÂÖ•Áî®Êà∑ÂêçÔºÅ");
        ui->usernameEdit->setFocus();
        return;
    }
    
    if (password.isEmpty()) {
        QMessageBox::warning(this, "ÁôªÂΩïÂ§±Ë¥•", "ËØ∑ËæìÂÖ•ÂØÜÁ†ÅÔºÅ");
        ui->passwordEdit->setFocus();
        return;
    }
    
    if (validateCredentials(username, password)) {
        QMessageBox::information(this, "ÁôªÂΩïÊàêÂäü", 
            QString("Ê¨¢ËøéÊÇ®Ôºå%1ÔºÅ\nÁôªÂΩïÊó∂Èó¥Ôºö%2")
            .arg(username)
            .arg(QDateTime::currentDateTime().toString("yyyy-MM-dd hh:mm:ss")));
        accept();
    } else {
        QMessageBox::critical(this, "ÁôªÂΩïÂ§±Ë¥•", 
            "Áî®Êà∑ÂêçÊàñÂØÜÁ†ÅÈîôËØØÔºÅ\n\nÊèêÁ§∫ÔºöÈªòËÆ§Ë¥¶Êà∑\nÁî®Êà∑ÂêçÔºöadmin ÂØÜÁ†ÅÔºöadmin123\nÁî®Êà∑ÂêçÔºöuser ÂØÜÁ†ÅÔºöpassword");
        ui->passwordEdit->clear();
        ui->passwordEdit->setFocus();
    }
}

void LoginDialog::onResetClicked()
{
    ui->usernameEdit->clear();
    ui->passwordEdit->clear();
    ui->usernameEdit->setFocus();
}

void LoginDialog::onRegisterClicked()
{
    RegisterDialog *registerDialog = new RegisterDialog(this);
    if (registerDialog->exec() == QDialog::Accepted) {
        QMessageBox::information(this, "ÊèêÁ§∫", "Ê≥®ÂÜåÊàêÂäüÔºÅÊÇ®Áé∞Âú®ÂèØ‰ª•‰ΩøÁî®Êñ∞Ë¥¶Êà∑ÁôªÂΩï„ÄÇ");
    }
    delete registerDialog;
}

void LoginDialog::togglePasswordVisibility()
{
    if (ui->togglePasswordButton->isChecked()) {
        ui->passwordEdit->setEchoMode(QLineEdit::Normal);
        ui->togglePasswordButton->setText("üôà");
    } else {
        ui->passwordEdit->setEchoMode(QLineEdit::Password);
        ui->togglePasswordButton->setText("üëÅ");
    }
}

bool LoginDialog::validateCredentials(const QString &username, const QString &password)
{
    QFile file("users.txt");
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        qDebug() << "Cannot open users.txt";
        return false;
    }
    
    QTextStream in(&file);
    QString hashedPassword = hashPassword(password);
    
    qDebug() << "ÁôªÂΩïÂ∞ùËØï - Áî®Êà∑:" << username << "ÂìàÂ∏å:" << hashedPassword;
    
    while (!in.atEnd()) {
        QString line = in.readLine();
        QStringList parts = line.split(":");
        qDebug() << "Ê£ÄÊü•Ë°å:" << line;
        if (parts.size() == 2 && parts[0] == username && parts[1] == hashedPassword) {
            qDebug() << "ÁôªÂΩïÈ™åËØÅÊàêÂäü!";
            return true;
        }
    }
    
    qDebug() << "ÁôªÂΩïÈ™åËØÅÂ§±Ë¥•!";
    return false;
}

void LoginDialog::saveUserCredentials(const QString &username, const QString &password)
{
    QFile file("users.txt");
    if (!file.open(QIODevice::WriteOnly | QIODevice::Append | QIODevice::Text)) {
        qDebug() << "Cannot open users.txt for writing";
        return;
    }
    
    QTextStream out(&file);
    out << username << ":" << hashPassword(password) << "\n";
}

QString LoginDialog::hashPassword(const QString &password)
{
    return QCryptographicHash::hash(password.toUtf8(), QCryptographicHash::Sha256).toHex();
}